types/webmidi/index.d.ts

```diff
@@ -4,193 +4,234 @@
+
+        /**
+         * The state of the device.
+         */
+        state: MIDIPortDeviceState;
+
+        /**
+         * The state of the connection to the device.
+         */
+        connection: MIDIPortConnectionState;
+
+        /**
+         * The handler called when an existing port changes its state or connection
+         * attributes.
+         */
+        onstatechange(e: MIDIConnectionEvent): void;
+
+        addEventListener(
+            type: 'statechange',
+            listener: (this: this, e: MIDIConnectionEvent) => any,
+            options?: boolean | AddEventListenerOptions,
+        ): void;
+        addEventListener(
+            type: string,
+            listener: EventListenerOrEventListenerObject,
+            options?: boolean | AddEventListenerOptions,
+        ): void;
+
+        /**
+         * Makes the MIDI device corresponding to the MIDIPort explicitly available. Note
+         * that this call is NOT required in order to use the MIDIPort - calling send() on
+         * a MIDIOutput or attaching a MIDIMessageEvent handler on a MIDIInputPort will
+         * cause an implicit open().
+         *
+         * When invoked, this method returns a Promise object representing a request for
+         * access to the given MIDI port on the user's system.
+         */
+        open(): Promise<MIDIPort>;
+
+        /**
+         * Makes the MIDI device corresponding to the MIDIPort
+         * explicitly unavailable (subsequently changing the state from "open" to
+         * "connected"). Note that successful invocation of this method will result in MIDI
+         * messages no longer being delivered to MIDIMessageEvent handlers on a
+         * MIDIInputPort (although setting a new handler will cause an implicit open()).
+         *
+         * When invoked, this method returns a Promise object representing a request for
+         * access to the given MIDI port on the user's system. When the port has been
+         * closed (and therefore, in exclusive access systems, the port is available to
+         * other applications), the vended Promise is resolved. If the port is
+         * disconnected, the Promise is rejected.
+         */
+        close(): Promise<MIDIPort>;
+    }
+
+    interface MIDIInput extends MIDIPort {
+        type: 'input';
+        onmidimessage(e: MIDIMessageEvent): void;
+
+        addEventListener(
+            type: 'midimessage',
+            listener: (this: this, e: MIDIMessageEvent) => any,
+            options?: boolean | AddEventListenerOptions,
+        ): void;
+        addEventListener(
+            type: 'statechange',
+            listener: (this: this, e: MIDIConnectionEvent) => any,
+            options?: boolean | AddEventListenerOptions,
+        ): void;
+        addEventListener(
+            type: string,
+            listener: EventListenerOrEventListenerObject,
+            options?: boolean | AddEventListenerOptions,
+        ): void;
+    }
```

Comment:
Same Q as above re: necessity of redefining these (I genuinely don't know -- if TS requires all of them to be redeclared every time this makes sense).